backend.go

```diff
@@ -87,81 +94,223 @@ func (b *azureSecretBackend) periodicFunc(ctx context.Context, sys *logical.Requ
+		return nil
+	}
+
+	b.Logger().Debug("retrieving service principal", "id", role.ServicePrincipalID)
+	servicePrincipal, err := client.getSP(ctx, role.ServicePrincipalID)
+	if err != nil {
+		return fmt.Errorf("error getting service principal: %w", err)
+	}
+
+	// we don't have the keyID and the service principal has a single password that hasn't expired, so we assume that is the
+	// password we are currently using and update the role.
+	if role.Credentials.KeyId == "" && len(servicePrincipal.PasswordCredentials) == 1 && servicePrincipal.PasswordCredentials[0].EndDate.After(expirationCutoff) {
+		cred := servicePrincipal.PasswordCredentials[0]
+		role.Credentials.KeyId = *cred.KeyID
+		role.Credentials.ExpiresAt = cred.EndDate.ToTime()
+
+		err = saveRole(ctx, storage, role, roleName)
+		if err != nil {
+			return fmt.Errorf("failed to save role %q: %w", roleName, err)
 		}
+		b.Logger().Debug("updated role credentials", "role", roleName)
+		return nil
+	}
 
-		b.Logger().Debug("periodic func", "rotate-root", "updating config with new password")
-		config.ClientSecret = config.NewClientSecret
-		config.ClientSecretKeyID = config.NewClientSecretKeyID
-		config.RootPasswordExpirationDate = config.NewClientSecretExpirationDate
-		config.NewClientSecret = ""
-		config.NewClientSecretKeyID = ""
-		config.NewClientSecretCreated = time.Time{}
+	currPassword, ok := findPasswordWithID(role.Credentials.KeyId, servicePrincipal.PasswordCredentials)
+	if !ok || currPassword.EndDate.Before(expirationCutoff) {
+		// we don't know what password we are using or it is expiring so refresh it
 
-		err = b.saveConfig(ctx, config, sys.Storage)
+		newPassword, err := client.addSPPassword(ctx, role.ServicePrincipalID, spExpiration)
 		if err != nil {
-			return err
+			return fmt.Errorf("failed to add new service principal password for role %q: %w", roleName, err)
 		}
+		b.Logger().Debug("added new service principal credentials", "role", roleName)
 
-		b.updatePassword = false
+		role.Credentials = &ClientCredentials{
+			KeyId:     *newPassword.KeyID,
+			Password:  *newPassword.SecretText,
+			ExpiresAt: newPassword.EndDate.ToTime(),
+		}
+		err = saveRole(ctx, storage, role, roleName)
+		if err != nil {
+			// try to remove the new password
+			err = client.removeSPPassword(ctx, role.ServicePrincipalID, *newPassword.KeyID)
+			if err != nil {
+				b.Logger().Warn("failed to remove new password for role", "role", roleName, "error", err)
+			}
+			return fmt.Errorf("failed to save role %q: %w", roleName, err)
+		}
+		b.Logger().Debug("updated role credentials", "role", roleName)
+
+		// remove all expired passwords
+		now := time.Now()
+		for _, pw := range servicePrincipal.PasswordCredentials {
+			if pw.EndDate.Before(now) {
+				err = client.removeSPPassword(ctx, role.ServicePrincipalID, *pw.KeyID)
+				if err != nil {
+					b.Logger().Warn("failed to remove old password for role", "role", roleName, "error", err)
+				}
+			}
+		}
```

Comment:
when do we expect this to trigger? if the expiration is in the future, but before the `expirationCutoff`, we will not enter this code, until we are ~+2y in the future?

Reply:
yeah, I think that is fine. In that case then each SP will only have at most 2 passwords at a given time. This is mostly to ensure we don't just build up passwords endlessly.